allow file:/// resources

clburlison opened the pull request:
Allow support for file:/// type resources. This was tested with dmgs, packages and the configuration plist so we should be good. 

Was too simple for me to not make this change earlier.

grahamgilbert commented:
Good so far. Rather than printing your errors to stdout, would you be able to get them displaying like the other imagr errors?

clburlison commented:
Alright. I can work on that. I also found out I didn't actually test packages properly. So additional changes are needed.

grahamgilbert commented:
Excellent :)

Sent from my iPhone




On Thu, Feb 25, 2016 at 10:15 AM -0800, "Clayton Burlison" <notifications@github.com> wrote:












Alright. I can work on that. I also found out I didn't actually test packages properly. So additional changes are needed. 



—
Reply to this email directly or view it on GitHub.

clburlison committed:
pacakages & dmgs now work with file resources

additional tests for script "type" need to be ran

Commit:
download scripts from file:/// or http://

clburlison commented:
Alright, if I didn't screw anything up this is good to go. I'm sending error messages for packages/dmgs to the Imagr error message screen. As it stands scripts are using the generic, "Couldn't copy script XX" message. I wasn't 100% sure how to tie the http error codes in for that section.

If anyone wants to give this a spin here is a sample imagr_config.plist

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>workflows</key>
    <array>
      <dict>
        <key>name</key>
        <string>Test install</string>
        <key>restart_action</key>
        <string>none</string>
        <key>description</key>
        <string>This is will install stuff from a file:/// resource</string>
        <key>components</key>
        <array>
          <dict>
            <key>type</key>
            <string>image</string>
            <key>url</key>
            <string>file:///Users/Shared/imagr/images/osx_updated_160129-10.11.3-15D21.hfs.dmg</string>
          </dict>
          <dict>
            <key>type</key>
            <string>package</string>
            <key>url</key>
            <string>file:///Users/Shared/imagr/cltools/Command_Line_Tools_OS_X_10.11_for_Xcode_7.2.dmg</string>
            <key>first_boot</key>
            <false/>
          </dict>
          <dict>
            <key>type</key>
            <string>package</string>
            <key>url</key>
            <string>file:///Users/Shared/imagr/packages/admin_user.pkg</string>
            <key>first_boot</key>
            <true/>
          </dict>
          <dict>
            <key>type</key>
            <string>script</string>
            <key>url</key>
            <string>file:///Users/Shared/imagr/scripts/awesomeness.sh</string>
            <key>first_boot</key>
            <false/>
          </dict>
          <dict>
            <key>type</key>
            <string>script</string>
            <key>content</key>
            <string>#!/bin/bash
  echo "&lt;"
  echo "{{target_volume}}"
  /usr/bin/touch "{{target_volume}}/some_file"</string>
          </dict>
        </array>
      </dict>
    </array>
  </dict>
</plist>
```

Working on documentation next.

grahamgilbert commented:
Nice one, I'll try and have a play tomorrow.

grahamgilbert commented on Imagr/Utils.py:
```diff
@@ -517,14 +536,29 @@ def unmountdmg(mountpoint):
+            return source, None
+        except:
+            error = "Unable to copy %s" % url
+            return None, error
```
Shouldn't this return False?

Commit:
return false if we can't find the file

clburlison commented:
So I'm running into an issue with COSXIP and local files. Need to find physical hardware to see why the installer is failing. I started the documentation and can merge whenever: https://github.com/clburlison/imagr/wiki/Creating-an-External-Drive

grahamgilbert commented:
I was going to suggest hdiutil wasn't liking the `file://` url, but I just tested that and it mounted fine. Sorry, can't be of more use.

clburlison commented:
I'm an idiot. All is well. I was attempting to escape a path and python wouldn't give me a good error message until I tired it on a physical machine. 

As far as I am concerned this PR is ready to be merged. Lets hold off a few days as I've got two people testing it (or they say they are).

I will need help with merging the Creating-an-External-Drive doc to the wiki as it has pictures embed. The only way GH wiki will embed pictures is with a git push so you'd need to do the merge yourself or give me temporary commit access to this repo.

grahamgilbert commented:
Add the pictures in somewhere with this PR and I'll merge it before the release next week.

Commit:
add wiki images

Commit:
Merge branch 'master' of https://github.com/grahamgilbert/imagr into allow_local_files

* 'master' of https://github.com/grahamgilbert/imagr:
  Allow for using a local dmg download of Imagr for building the NBI
  Add dimensions of screen to background image request
  Validate URLs.
  Verify value must be a bool.
  remove whitespace
  validateplist from tempfile always
  Add verify key to image tasks.
  Fix color format for network status.
  Fix applications launching behind Imagr window.
  Fix window ordering regression.
  Send OS version and build with background image request.
  Display background image.
  Add background window.
  Fix python format string.
  Fix deployment target warning.
  Add color to first-boot messages.
  Update to LoginLog 1.1.
  validateplist-plistlib

Commit:
update validateplist for local files

Commit:
slight update for downloadChunks

Due to how Imagr is passing to `installer` and `hdiutil` we can't treat all file:/// resources the same. As such, we mandate that all local resources have spaces escaped as %20 then manually remove the %20 when passing to `installer` and when copying the files via shutil.copy.

Commit:
encode space characters as %20

Commit:
update wiki images for NBICreator

clburlison commented:
- Images have been added to this commit.
- [Documentation updated](https://github.com/clburlison/imagr/wiki/Creating-an-External-Drive) (I'll update the wiki after this is merged).
- Validateplist updated for file:/// urls. 
- I made one more minor change to how we're dealing with URLs. This fixes an issue with spaces in the URL. https://github.com/grahamgilbert/imagr/pull/128/commits/31f860f7514c5f22dc7d943161b3d415af097a60

> Due to how Imagr is passing to `installer` and `hdiutil` we can't treat all file:/// resources the same. As such, we mandate that all local resources have spaces escaped as %20 then manually remove the %20 when passing to `installer` and when copying the files via `shutil.copy`.

The above commit really is only for NBICreator as it forces you to use "Image Volume" as the name of your external drive when booted into the NetInstall Environment. A slight limitation but pretty easy to work around.

I've also passed all tests I could think of.

grahamgilbert commented:
Ok, let's do this thing. Merging.
